tools/indexer/cmd/start.go

```diff
@@ -0,0 +1,284 @@
+				initLogger.Error("[SHUTDOWN] failed to save last block number", "error", err)
+			}
+			initLogger.Info("[SHUTDOWN] indexer stopped", "block", lastBN)
+			return nil
+
+		case <-mainTicker.C:
+			nextBN := lastBN + 1
+
+			// Fetch execution block data
+			ei, err := beacon.FetchCombinedBlockData(ctx, httpc, infuraRPC, beaconBase, nextBN)
+			if err != nil || ei == nil {
+				initLogger.Warn("[BLOCK] not available yet", "block", nextBN, "error", err)
+				continue
+			}
+			fields := []any{
+				"block", nextBN,
+				"slot", ei.Slot,
+			}
+			if ei.Timestamp != nil {
+				fields = append(fields, "timestamp", ei.Timestamp.Format(time.RFC3339))
+			}
+			if ei.ProposerIdx != nil {
+				fields = append(fields, "proposer_index", *ei.ProposerIdx)
+			}
+			if ei.RelayTag != nil {
+				fields = append(fields, "winning_relay", *ei.RelayTag)
+			}
+			if ei.BuilderHex != nil {
+				pref := *ei.BuilderHex
+				if len(pref) > 20 {
+					pref = pref[:20]
+				}
+				fields = append(fields, "builder_pubkey_prefix", pref)
+			}
+			if ei.RewardEth != nil {
+				fields = append(fields, "producer_reward_eth", *ei.RewardEth)
+			}
+			initLogger.Info("processing block", fields...)
+
+			// Save block to database
+			if err := db.UpsertBlockFromExec(ctx, ei); err != nil {
+				initLogger.Error("[DB] failed to save block", "block", nextBN, "error", err)
+				continue
+			}
+			initLogger.Info("[DB] block saved successfully", "block", nextBN)
+
+			// Fetch and store bid data from all relays
+			totalBids := 0
+			successfulRelays := 0
+			mainContextCanceled := false
+			const batchSize = 500
+			for _, rr := range relays {
+				// Check if main context is canceled before processing each relay
+				if ctx.Err() != nil {
+					initLogger.Warn("main context canceled, stopping relay processing")
+					mainContextCanceled = true
+					break
+				}
+
+				bids, err := relay.FetchBuilderBlocksReceived(ctx, httpc, rr.URL, ei.Slot)
+				if err != nil {
+					initLogger.Error("[RELAY] failed to fetch bids", "relay_id", rr.ID, "url", rr.URL, "error", err)
+					continue
+				}
+
+				relayBids := 0
+				batch := make([]database.BidRow, 0, batchSize)
+
+				for _, bid := range bids {
+					// Check if main context is still valid
+					if ctx.Err() != nil {
+						initLogger.Warn("[BIDS] main context canceled, stopping bid insertion")
+						mainContextCanceled = true
+						break
+					}
+
+					if row, ok := relay.BuildBidInsert(ei.Slot, rr.ID, bid); ok {
+						batch = append(batch, row)
+
+						if len(batch) >= batchSize {
+							insCtx, cancel := context.WithTimeout(ctx, 5*time.Second)
+							if err := db.InsertBidsBatch(insCtx, batch); err != nil {
+
+								initLogger.Error("[DB]batch insert failed", "slot", ei.Slot, "relay_id", rr.ID, "count", len(batch), "error", err)
+							} else {
+								relayBids += len(batch)
+							}
+							cancel()
+							batch = batch[:0]
+						}
+					}
+				}
+
+				// final flush
+				if len(batch) > 0 {
```

Comment:
If main context is cancelled, this insert will be called with cancelled context.